### PR TITLE
feat: Add a utility to identify tag editor tags states

### PR DESCRIPTION
**Pull request by taheramr**
### Description

This PR creates a utility to return back the tags states (created, removed or updated) inside the <TagEditor /> component.

This helper function can be used instead of customers implementing their own diffing algorithm.

TT -> P154609808

### How has this been tested?
Added unit tests for the utility.

<details>
   <summary>Review checklist</summary>

_The following items are to be evaluated by the author(s) and the reviewer(s)._

#### Correctness

- _Changes include appropriate documentation updates._
- _Changes are backward-compatible if not indicated, see [`CONTRIBUTING.md`](https://github.com/cloudscape-design/components/blob/main/CONTRIBUTING.md#public-apis)._
- _Changes do not include unsupported browser features, see [`CONTRIBUTING.md`](https://github.com/cloudscape-design/components/blob/main/CONTRIBUTING.md#browsers-support)._
- _Changes were manually tested for accessibility, see [accessibility guidelines](https://cloudscape.design/foundation/core-principles/accessibility/)._

#### Security

- _If the code handles URLs: all URLs are validated through [the `checkSafeUrl` function](https://github.com/cloudscape-design/components/blob/main/src/internal/utils/check-safe-url.ts)._

#### Testing

- _Changes are covered with new/existing unit tests?_
- _Changes are covered with new/existing integration tests?_
</details>

---

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

### CODE REVIEW

File: src/tag-editor/utils.ts [hunk 1]
```diff
@@ -46,3 +50,46 @@ export function useMemoizedArray<T>(array: ReadonlyArray<T>, isEqual: (prev: T,
   }, [updated]);
   return updated;
 }
+
+interface GetTagsDiffResult {
+  createdTags: Record<string, string>;
```
Comment (georgylobko) [inline src/tag-editor/utils.ts:55]: [minor] I would suggest to rename these props to just `created` and `removed`, since the function name already has `tags` in the name

**Review by georgylobko — DISMISSED** [review]

File: src/tag-editor/__tests__/get-tags-diff.test.tsx [hunk 2]
```diff
@@ -0,0 +1,81 @@
+// Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
+// SPDX-License-Identifier: Apache-2.0
+import { warnOnce } from '@cloudscape-design/component-toolkit/internal';
+
+import { TagEditorProps } from '../../../lib/components/tag-editor';
+import { getTagsDiff } from '../utils';
+
+jest.mock('@cloudscape-design/component-toolkit/internal', () => ({
+  warnOnce: jest.fn(),
+}));
```
Comment (just-boris) [inline src/tag-editor/__tests__/get-tags-diff.test.tsx:10]: This mocks too much, all of the toolkit internals. Even if your tests pass, you make live for future devs editing this code harder.

The intended setup for testing warnings is here: https://github.com/cloudscape-design/components/blob/7f14eff7001437b54eeeca2ccf6c165afc56313d/src/app-layout/__tests__/multi-layout.test.tsx#L192-L199

File: src/tag-editor/__tests__/get-tags-diff.test.tsx [hunk 3]
```diff
@@ -0,0 +1,81 @@
+// Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
+// SPDX-License-Identifier: Apache-2.0
+import { warnOnce } from '@cloudscape-design/component-toolkit/internal';
+
+import { TagEditorProps } from '../../../lib/components/tag-editor';
+import { getTagsDiff } from '../utils';
```
Comment (just-boris) [inline src/tag-editor/__tests__/get-tags-diff.test.tsx:6]: Both imports should come from the same source, either the `src` or `lib` folder, but not mixed

File: src/tag-editor/__tests__/get-tags-diff.test.tsx [hunk 4]
```diff
@@ -0,0 +1,81 @@
...
+
+  it('should warn if initial tags are missing the `existing` property', () => {
+    const invalidInitialTags: TagEditorProps.Tag[] = [{ key: 'tag1', value: 'value1', existing: false }];
+
+    getTagsDiff(invalidInitialTags, initialTags);
+
+    expect(warnOnce).toHaveBeenCalledWith(
+      'identifyTagStates',
+      'all initial tags should have `existing` property set to `true`.'
+    );
```
Comment (just-boris) [inline src/tag-editor/__tests__/get-tags-diff.test.tsx:72]: Fun fact, because you do not reset the mock state properly, this check will pass even if you put it into the next test.

The recommended mock setup from above addresses this issue for you, but you need to understand what is happening here

File: src/tag-editor/__tests__/get-tags-diff.test.tsx [hunk 5]
```diff
@@ -0,0 +1,81 @@
...
+import { warnOnce } from '@cloudscape-design/component-toolkit/internal';
+
+import { TagEditorProps } from '../../../lib/components/tag-editor';
+import { getTagsDiff } from '../utils';
+
+jest.mock('@cloudscape-design/component-toolkit/internal', () => ({
+  warnOnce: jest.fn(),
+}));
+
+describe('identifyTagStates', () => {
```
Comment (just-boris) [inline src/tag-editor/__tests__/get-tags-diff.test.tsx:12]: To make sure we cover all corner cases, we should have "all tags created" test case `getTagsDiff({}, newTags)` and "all tags removed" – `getTagsDiff(initialTags, {})`

File: src/tag-editor/utils.ts [hunk 6]
```diff
@@ -46,3 +50,46 @@ export function useMemoizedArray<T>(array: ReadonlyArray<T>, isEqual: (prev: T,
...
+ * the removed and added tags.
+ *
+ * @param initialTags - The original tags fetched from the backend or tagging service.
+ * @param tags - The current tags provided to the tag editor, including any new or modified tags.
+ * @returns An object containing two arrays:
+ * - `created`: An record of tags that are new or updated (with modified values).
+ *    Each tag is represented by its `key` and `value`.
+ * - `removed`: An array of tag keys that were present in the initial tags but marked for removal.
+ */
+
```
Comment (just-boris) [inline src/tag-editor/utils.ts:73]: Why is there an empty line? Usually it disconnects JSDOC from the function

File: src/tag-editor/utils.ts [hunk 7]
```diff
@@ -46,3 +50,46 @@ export function useMemoizedArray<T>(array: ReadonlyArray<T>, isEqual: (prev: T,
...
+  }
+  const updated = tags.filter(tag =>
+    initialTags.some(({ key, value }) => {
+      return !tag.markedForRemoval && tag.key === key && tag.existing && tag.value !== value;
+    })
+  );
+  const created = [...tags.filter(tag => !tag.existing), ...updated]
+    .map(({ key, value }) => ({ key, value }))
+    .reduce((acc: Record<string, string>, tag) => {
+      acc[tag.key] = tag.value;
```
Comment (just-boris) [inline src/tag-editor/utils.ts:89]: Should we print a warning, if there are duplicate keys?

File: src/tag-editor/__tests__/get-tags-diff.test.tsx (hunk 5 above)
Comment (taheramr) [reply src/tag-editor/__tests__/get-tags-diff.test.tsx:7]: good call!

added these tests

File: src/tag-editor/utils.ts (hunk 7 above)
Comment (taheramr) [reply src/tag-editor/utils.ts:102]: Added that logic and also added it behind the `isDevelopment` flag

File: src/tag-editor/utils.ts (hunk 1 above)
Comment (taheramr) [reply src/tag-editor/utils.ts]: updated - good call

File: src/tag-editor/__tests__/get-tags-diff.test.tsx [hunk 8]
```diff
@@ -0,0 +1,97 @@
...
+    expect(result.removed).toEqual(['tag3', 'tag1']);
+  });
+
+  test('should return empty arrays if no changes are detected', () => {
+    const result = getTagsDiff(initialTags, initialTags);
+
+    expect(result.created).toEqual({});
+    expect(result.removed).toEqual([]);
+  });
+
```
Comment (georgylobko) [inline src/tag-editor/__tests__/get-tags-diff.test.tsx:63]: Let's add an additional test case when `initialTags` and `tags` are present and equal

File: src/tag-editor/__tests__/get-tags-diff.test.tsx (hunk 5 above)
Comment (just-boris) [reply src/tag-editor/__tests__/get-tags-diff.test.tsx:7]: Still do not see this test case: `getTagsDiff(initialTags, [])`

File: src/tag-editor/utils.ts (hunk 7 above)
Comment (just-boris) [reply src/tag-editor/utils.ts:102]: `warnOnce` already has `isDevelopment` check built-in, but an extra one is fine too

File: src/tag-editor/__tests__/get-tags-diff.test.tsx (hunk 8 above)
Comment (taheramr) [reply src/tag-editor/__tests__/get-tags-diff.test.tsx:68]: done

**Review by just-boris — COMMENTED** [review]
all good for the code, but let's add all suggested tests from me and Georgii

File: src/tag-editor/utils.ts (hunk 7 above)
Comment (taheramr) [reply src/tag-editor/utils.ts:102]: yes but I don't want to execute the checking logic in production when its not gonna warn or do anything anyways..

**Review by just-boris — APPROVED** [review]

**Review by georgylobko — APPROVED** [review]
